shell/platform/windows/windows_lifecycle_manager.cc

```diff
@@ -75,6 +75,24 @@ bool WindowsLifecycleManager::WindowProc(HWND hwnd,
     case WM_DWMCOMPOSITIONCHANGED:
       engine_->OnDwmCompositionChanged();
       break;
+
+    case WM_SIZE:
```

Comment:
The app can have native (non-Flutter) windows. Alternatively, the app can have two separate engines each with their own window.

Should this logic be inspecting these other windows when generating app lifecycle events?

Reply:
I am not aware of any way to respond to window notifications in non-flutter windows other than by expecting the developer to make a call to our API from their window's WndProc. In what case does the app have two engines?

Reply:
> In what case does the app have two engines?

Currently that's the only way to do a multi-window app using Flutter.

> I am not aware of any way to respond to window notifications in non-flutter windows other than by expecting the developer to make a call to our API from their window's WndProc

For the application exist lifecycle event, the app looks up all other windows and examines their state. Can we do similar tricks for hidden/shown state?

Taking a step back, what are the uses experiences we're hoping to enable with this change? It's not clear to me that sending potentially incorrect application lifecycle events is better than not sending application lifecycle events.

Reply:
The use case I'm thinking of are an application that wants to pause an operation when no windows are visible.  The framework also pauses frame generation when no windows are visible (see [here](https://github.com/gspencergoog/flutter/blob/link_examples/packages/flutter/lib/src/scheduler/binding.dart#L404).

Reply:
So, I can see an app wanting to sort of hibernate when no windows (including non-Flutter) windows are visible.

Mostly, I want the concept to be simple to understand, with as few caveats as possible: e.g. "The app will be in the 'hidden' state when zero of your application windows are visible to the user.", rather than having to qualify/complicate that concept with "just Flutter views (Additional platform windows don't count)".

Reply:
So are you suggesting that we should not merge this PR until we have a way to respond to changes in non-Flutter windows' states?

Reply:
Yes, like we talked about in the meeting: Loïc pointed out that we don't want to be sending incorrect events, and he's right. @loic-sharma you had ideas for how to listen to the app-wide events, right?

If it's not possible to listen to the events on other windows, then the backup plan would be to provide an embedder API that allows an app to tell us when all the non-Flutter windows lose or gain focus or visibility. Probably we would want to just let them indicate how many are visible and if any are focused.

Reply:
Well I'm trying to use Spy++, but when opening a log for a Window it doesn't actually log any messages, so that's not very helpful.

Anyway, from experience I've observed that the top level window receives SETFOCUS when it first appears and never receives KILLFOCUS, and that while the top level window receives SHOWWINDOW events, the Flutter window does not. Note that even the MS docs state for some paired messages that a window may receive one and not the other.

This means we cannot just process all messages for the process in the same way. I think providing an API function that a developer can call from their WndProc may be preferable anyway, because that will allow a developer to exclude certain windows from the lifecycle logic if they so choose, and to implement custom logic for focus or visibility if necessary

Reply:
> @loic-sharma you had ideas for how to listen to the app-wide events, right?

The idea I suggested was to listen to the [Windows message loop](https://github.com/flutter/flutter/blob/ef43463c86bf52fbb5906a4526188b04570f73a0/packages/flutter_tools/templates/app_shared/windows.tmpl/runner/main.cpp.tmpl#L35-L39). When we receive a message that indicates a window was minimized/shown, we [enumerate the app's windows](https://github.com/flutter/engine/blob/4b274b5c2c3a3cd6e9e04041fea730952605ac57/shell/platform/windows/windows_lifecycle_manager.cc#L126-L135) and check their visibility. If a window was minimized and all windows are minimized, we send the `AppLifecycleState.hidden` state. If a window was shown and it is the only visible window, we send the `AppLifecycleState.resumed` state.

> I think providing an API function that a developer can call from their WndProc may be preferable anyway, because that will allow a developer to exclude certain windows from the lifecycle logic if they so choose, and to implement custom logic for focus or visibility if necessary

This is a viable alternative too.

### Comparison

Option 1 - introduce APIs that the developer can call from their window procedure:
* ✅ Works on an app where all windows are created by Flutter
* ⚠️ Requires modifications to the runner app template
* ⚠️ Plugins that create windows - if there are any - will need to be updated
* ⚠️ Flutter will send incorrect lifecycle events if:
    * A developer uses win32 APIs to create a window but forgets to notify Flutter
    * A developer uses code they do not control to create a window. For example, the app uses a win32 API which creates a new window whose code - including the window procedure - is entirely owned by Windows
 
Option 2 - listen to the windows event loop
* ✅ Works on an app where all windows are created by Flutter
* ⚠️ Requires modifications to the runner app template
* ✅ Plugins that create windows - if there are any - will work
* ⚠️ This might be trickier to implement properly

Reply:
What modifications to the template are you envisioning us needing to make for option 1? As I see it, I do not see why we would need to do so.

Reply:
I suspect option 2 could also introduce some other issues if/when a window handles visibility/focus differently than default, like our top level runner window. Also, there's an issue of whether we listen to messages on all windows or only top level windows. In the former case, win32 controls like text edits and buttons will count as visible windows to keep track of, and in the latter, the FlutterView itself is excluded.

Reply:
Agreed. If we go with option 1 we should:

1. Document how to notify Flutter of external windows
2. Determine which Windows plugins, if any, create windows. We should work with their maintainers to ensure it does not break Flutter's app lifecycle events

Also, we should make sure the API supports this scenario where the developer cannot modify a window's procedure:
> A developer uses code they do not control to create a window. For example, the app uses a win32 API which creates a new window whose code - including the window procedure - is entirely owned by Windows

Reply:
Those first two should be doable. Modifying an existing window's WndProc is probably possible (e.g. see https://stackoverflow.com/questions/63143237/change-wndproc-of-the-window). Only thing is we would need a place to store a pointer to the original WndProc to call from the replacement one. We could add a `map<HWND, WNDPROC>` somewhere, perhaps in `FlutterWindowsEngine`. Any thoughts?

Reply:
I think everything in this thread is addressed as we decided not to modify external window's procedures. Feel free to mark this as resolved :)